slither/detectors/functions/external_function.py

```diff
@@ -105,9 +107,31 @@ def get_all_function_definitions(base_most_function):
     def function_parameters_written(function):
         return any(p in function.variables_written for p in function.parameters)
 
+    @staticmethod
+    def is_reference_type(parameter):
+        if isinstance(parameter.type, ArrayType):
+            return True
+        if isinstance(parameter.type, MappingType):
+            return True
+        if isinstance(parameter.type, UserDefinedType) and isinstance(
+            parameter.type.type, Structure
+        ):
+            return True
+        if str(parameter.type) in ["bytes", "string"]:
+            return True
+        return False
+
     def _detect(self):  # pylint: disable=too-many-locals,too-many-branches
         results = []
 
+        # After solc 0.6.9, calldata arguments are allowed in public functions
+        if self.compilation_unit.solc_version >= "0.7." or self.compilation_unit.solc_version in [
```

Comment:
I think there likely needs to be a lower bound since data location doesn't exist in 0.4.25, for instance.

Reply:
If I understood correctly, it does [exist](https://docs.soliditylang.org/en/v0.4.25/types.html#:~:text=There%20is%20also%20a%20third%20data%20location%2C%20calldata%2C%20which%20is%20a%20non%2Dmodifiable%2C%20non%2Dpersistent%20area%20where%20function%20arguments%20are%20stored.%20Function%20parameters%20(not%20return%20parameters)%20of%20external%20functions%20are%20forced%20to%20calldata%20and%20behave%20mostly%20like%20memory.), but I think the `calldata` keyword was added later (probably in 0.5.0). 

So I think it would be nice to still detect if a function can be changed from external to public in these contexts, since this would make the compiler force the appropriate parameters to be located in calldata.

Reply:
You're correct